website/src/components/Visualisations/Visualisations.tsx

```diff
@@ -1,6 +1,154 @@
+  }, [])
+
+  const onDidMount = async () => {
+    setLoading({ loading: true, loadingText: `We're hard at work!` })
+    try {
+      const o = await loadLocalVisOutput()
+      if (o) {
+        setOutput(o)
+        history.push(gamevisualisation)
+      }
+    }
+    catch (err) {
+      // if error, just assume not stored at all
+      console.error(err)
+    }
+    setLoading(initialLoadingState)
+  }
+
+  useEffect(() => {
+    if (output) {
+      try {
+        const gotGitHash = output.GitInfo.Hash
+        if (gotGitHash !== GitHash) {
+          setWarning(`This website was built on commit "${GitHash}", and the output you're trying to visualise ` +
+            ` was produced on commit "${output.GitInfo.Hash}". There may be incompatibilities!`)
+        }
+        else {
+          setWarning(undefined)
+        }
+      }
+      catch (err) {
+        // can't read output.GitInfo.Hash, just reset which clears localforage as well
+        handleReset()
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [output])
+
+  const handleReset = async () => {
+    setLoading({ loading: true, loadingText: `Cleaning up your mess!` })
+    setOutput(undefined)
+    await clearLocalVisOutput()
+    history.push(visualisations)
+    setLoading(initialLoadingState)
+    setError(undefined)
+    setWarning(undefined)
+  }
+
+  const onUpload = async (event: React.ChangeEvent<HTMLInputElement>) => {
+    setLoading({ loading: true, loadingText: `Reading your file!` })
+
+    try {
+      // @ts-ignore silence, typechecker
+      const file: Blob | null = event.target.files[0]
+      if (!file || !(file instanceof Blob)) {
+        throw new Error('No or unsupported file uploaded')
+      }
+      const outputText = await file.text()
+      const o = JSON.parse(outputText) as OutputJSONType
+      try {
+        // find githash to check whether the JSON is ok
+        const gotGitHash = o.GitInfo.Hash
+        console.debug(gotGitHash)
+      } catch (err) {
+        throw new Error(`Unsupported file.`)
+      }
+      setOutput(o)
+      await storeLocalVisOutput(o)
+      setError(undefined)
+    }
+    catch (err) {
+      setError(err.message)
+    }
+    history.push(gamevisualisation)
+    setLoading(initialLoadingState)
+  }
+
+  return <>
+    {
+      output && <>
+        <VisualisationsNavbar reset={handleReset} />
+      </>
+    }
+    <div style={{ paddingTop: 24 }}>
+      {
+        error &&
+        <Alert variant="danger" onClose={() => setError(undefined)} dismissible className="custom-alert">
+          <Alert.Heading>Oh reeeeeeeeee!</Alert.Heading>
+          <p>{error}</p>
+        </Alert>
+      }
+      {
+        warning &&
+        <Alert variant="warning" onClose={() => setWarning(undefined)} dismissible className="custom-alert">
+          <Alert.Heading>Rough seas ahead!</Alert.Heading>
+          <p>{warning}</p>
+        </Alert>
+      }
+      {
+        output ?
+          <>
```

Comment:
Is a fragment needed if it only has 1 child?

Reply:
good spot